### PR TITLE
Job improvements

### PR DESCRIPTION
**Pull request by xabolcs**

### CODE REVIEW

**Comment by xabolcs** [comment]
Problem 1: doesn't compile due to Ebean leftovers :grinning: 

```
Compilation error (In /app/com/openseedbox/jobs/NodePollerJob.java around line 46)
The file /app/com/openseedbox/jobs/NodePollerJob.java could not be compiled. Error raised is : The method save() in the type ModelBase is not applicable for the arguments (List&lt;Torrent>)

play.exceptions.CompilationException: The method save() in the type ModelBase is not applicable for the arguments (List<Torrent>)
	at play.classloading.ApplicationCompiler$2.acceptResult(ApplicationCompiler.java:264)
	at org.eclipse.jdt.internal.compiler.Compiler.handleInternalException(Compiler.java:678)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:522)
	at play.classloading.ApplicationCompiler.compile(ApplicationCompiler.java:300)
	at play.classloading.ApplicationClassloader.getAllClasses(ApplicationClassloader.java:420)
	at play.Play.preCompile(Play.java:609)
	at play.Play.init(Play.java:309)
	at play.server.Server.main(Server.java:160)
```

**Comment by xabolcs** [comment]
Problem 2: NPE in the view :see_no_evil: 

```
frontend_1      | @7ppgnecd9
frontend_1      | Internal Server Error (500)
frontend_1      | Template execution error (In /app/views/admin/jobs.html around line 46)
frontend_1      | Execution error occurred in template /app/views/admin/jobs.html. Exception raised was NullPointerException : Cannot invoke method size() on null object.
frontend_1      | play.exceptions.TemplateExecutionException: Cannot invoke method size() on null object
frontend_1      | 	at play.templates.BaseTemplate.throwException(BaseTemplate.java:83)
frontend_1      | 	at play.templates.GroovyTemplate.internalRender(GroovyTemplate.java:280)
frontend_1      | 	at play.templates.Template.render(Template.java:26)
frontend_1      | 	at play.templates.GroovyTemplate.render(GroovyTemplate.java:203)
frontend_1      | 	at play.mvc.results.RenderTemplate.<init>(RenderTemplate.java:28)
frontend_1      | 	at play.mvc.Controller.renderTemplate(Controller.java:836)
frontend_1      | 	at play.mvc.Controller.renderTemplate(Controller.java:814)
frontend_1      | 	at play.mvc.Controller.render(Controller.java:876)
frontend_1      | 	at controllers.Admin.jobs(Admin.java:242)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at play.mvc.ActionInvoker.invokeWithContinuation(ActionInvoker.java:536)
frontend_1      | 	at play.mvc.ActionInvoker.invoke(ActionInvoker.java:473)
frontend_1      | 	at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:467)
frontend_1      | 	at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:436)
frontend_1      | 	at play.mvc.ActionInvoker.invoke(ActionInvoker.java:159)
frontend_1      | 	at play.server.ServletWrapper$ServletInvocation.execute(ServletWrapper.java:579)
frontend_1      | 	at play.Invoker$Invocation.run(Invoker.java:308)
frontend_1      | 	at play.server.ServletWrapper$ServletInvocation.run(ServletWrapper.java:570)
frontend_1      | 	at play.Invoker.invokeInThread(Invoker.java:70)
frontend_1      | 	at play.server.ServletWrapper.service(ServletWrapper.java:157)
frontend_1      | 	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:814)
frontend_1      | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:223)
frontend_1      | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
frontend_1      | 	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
frontend_1      | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)
frontend_1      | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
frontend_1      | 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177)
frontend_1      | 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
frontend_1      | 	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
frontend_1      | 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:119)
frontend_1      | 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
frontend_1      | 	at org.apache.catalina.valves.HealthCheckValve.invoke(HealthCheckValve.java:102)
frontend_1      | 	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
frontend_1      | 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
frontend_1      | 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)
frontend_1      | 	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:400)
frontend_1      | 	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
frontend_1      | 	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:859)
frontend_1      | 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1734)
frontend_1      | 	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
frontend_1      | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
frontend_1      | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
frontend_1      | 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
frontend_1      | 	at java.base/java.lang.Thread.run(Unknown Source)
frontend_1      | Caused by: java.lang.NullPointerException: Cannot invoke method size() on null object
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
frontend_1      | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
frontend_1      | 	at play.templates.GroovyTemplate.internalRender(GroovyTemplate.java:255)
frontend_1      | 	... 44 more
```

File: app/com/openseedbox/jobs/NodePollerJob.java [hunk 1]
```diff
@@ -38,7 +43,7 @@ protected Object doGenericJob() throws Exception {
 					db.merge(match);
 				}
 			}
-			Torrent.batch().update(fromDb);
+			Torrent.save(fromDb);
```
Comment (xabolcs) [inline app/com/openseedbox/jobs/NodePollerJob.java:46]: Simply roll back the unrelated change! 
```suggestion
			Torrent.batch().update(fromDb);
```

File: app/views/admin/jobs.html [hunk 2]
```diff
@@ -42,9 +42,27 @@
 		<p>No jobs have been run yet.</p>
 	#{/else}
 %{ } }%
+%{ def renderHeaderWithToc = { Map jobName, Map jobData ->
+	if (jobName.size() != jobData.size() || jobName.size == 0) {
```
Comment (xabolcs) [inline app/views/admin/jobs.html:46]: No need for `()` in `.size()`? Or an extra `?` is missing?

File: app/views/admin/jobs.html [hunk 3]
```diff
@@ -42,9 +42,27 @@
...
+	jobData.each { id, data ->
+		out.print('<h2 id="' + id + '">' + jobName[id] + '<small>&nbsp;<a href="' + play.mvc.Router.reverse('Admin.runJobManually', [type: id]).url + '">Run</a></small></h2>\n')
+		renderTable(data)
+	}
+}
+%}
 <span class="pull-right">Time: ${new Date()}</span>
-<h2>Poller Jobs <small><a href="@{Admin.runJobManually('poller')}">Run</a></small></h2>
-${renderTable(pollerJobs)}
+%{renderHeaderWithToc(jobNames, jobDatas) %}
```
Comment (xabolcs) [inline app/views/admin/jobs.html:65]: That NPE is caused by this change.
The `pollerJobs` -> `jobNames, jobDatas` is just the half of the whole change.

The change from `Admin.java` is missing here. :confused:

File: app/com/openseedbox/jobs/admin/MaintenanceJob.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Calendar;
+
+@Every("7d")
+@JobName("Weekly Maintenance Scheduler")
+public class  MaintenanceJob extends ErrorLoggedAdminJob {
+
+    @Override
+    protected Object doGenericJob() throws Exception {
+        new WorkerJob().run();
```
Comment (xabolcs) [inline app/com/openseedbox/jobs/admin/MaintenanceJob.java:16]: Running that worker job causes NullPointerException. :shrug: 

```java
Error during job execution (com.openseedbox.jobs.admin.MaintenanceJob)

Oops: NullPointerException
Unexpected error : Unexpected Error, caused by exception NullPointerException: null

play.exceptions.UnexpectedException: Unexpected Error
	at play.Invoker$Invocation.onException(Invoker.java:255)
	at play.jobs.Job.onException(Job.java:156)
	at play.jobs.Job.call(Job.java:233)
	at play.jobs.Job$1.call(Job.java:119)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.NullPointerException
	at play.jobs.JobsPlugin.afterInvocation(JobsPlugin.java:268)
	at play.plugins.PluginCollection.afterInvocation(PluginCollection.java:570)
	at play.Invoker$Invocation.after(Invoker.java:236)
	at play.jobs.Job.call(Job.java:229)
	... 7 more
```

Comment (xabolcs) [reply app/com/openseedbox/jobs/admin/MaintenanceJob.java]: Yeah ... `.run()` <> `.now()` :facepalm:

**Comment by xabolcs** [comment]
Interesting results.

Before this change, the node poller duration was < 100 ms.
Now its 200+ ms. :thinking: 

image

**Comment by xabolcs** [comment]
Hmmm .... looks like the database warmed itself. :grinning: 
All looks good now.

Screenshot 2022-12-11 at 20-08-55 OpenSeedbox - Admin

File: app/com/openseedbox/jobs/admin/WeeklyMaintenanceJob.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        new JobEventJob().now();
+        return null;
+    }
+
+    @JobName("JobEvent Maintenance")
+    public class JobEventJob extends LoggedAdminJob {
+        @Override
+        protected Object doGenericJob() throws Exception {
+            Calendar lastWeek = Calendar.getInstance();
+            lastWeek.add(Calendar.DAY_OF_YEAR, -7);
```
Comment (xabolcs) [inline app/com/openseedbox/jobs/admin/WeeklyMaintenanceJob.java:25]: It would be nice to externalize this to `openseedbox.maintenance.jobevent.weeks`, and force at least one week.

**Comment by xabolcs** [comment]
Weekly `JobEvent` Maintenance job looks good!

image

```
08:35:19,332 DEBUG ~ delete before date 2022-12-12T08:35:19.319Z
08:35:19,756 DEBUG ~ deleted 70917 rows
```

File: app/com/openseedbox/jobs/admin/WeeklyMaintenanceJob.java (hunk 5 above)
Comment (xabolcs) [reply app/com/openseedbox/jobs/admin/WeeklyMaintenanceJob.java]: - `openseedbox.maintenance.jobevent.older.than.weeks` and
- `openseedbox.maintenance.torrentevent.older.than.months`

**Comment by xabolcs** [comment]
Meh ... a regression like NPE :confused: 

STR:
* an active backend server is down (Tomcat up, but unhealty, because the Play! app is down)
* Admin / Nodes view shows correctly: "Error: Node returned status: 404. Probably java isnt running."
* Poller Jobs on Admin / Jobs shows an _almost_ correct trace
```
com.openseedbox.code.MessageException: Unsuccessful webservice call, status: 
	at com.openseedbox.models.Node.handleWebServiceResponse(Node.java:166)
	at com.openseedbox.backend.node.NodeBackend.getResponseBodyOrError(NodeBackend.java:275)
	at com.openseedbox.backend.node.NodeBackend.list(NodeBackend.java:242)
	at com.openseedbox.backend.node.NodeBackend.listTorrents(NodeBackend.java:226)
	at com.openseedbox.jobs.NodePollerJob$NodePollerWorker.doGenericJob(NodePollerJob.java:34)
	at com.openseedbox.jobs.GenericJob.runJob(GenericJob.java:21)
	at com.openseedbox.jobs.LoggedJob.runJob(LoggedJob.java:20)
	at com.openseedbox.jobs.GenericJob.doJobWithResult(GenericJob.java:15)
	at play.jobs.Job.call(Job.java:214)
	at play.jobs.Job$1.call(Job.java:119)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

But the log is spammed every for every job runs with:
```
10:35:55,101 ERROR ~ Oops. Error in Logger !
java.lang.NullPointerException
	at com.openseedbox.code.MessageException.getErrorDescription(MessageException.java:17)
	at play.Logger.niceThrowable(Logger.java:593)
	at play.Logger.error(Logger.java:453)
	at com.openseedbox.jobs.LoggedJob.runJob(LoggedJob.java:27)
	at com.openseedbox.jobs.GenericJob.doJobWithResult(GenericJob.java:15)
	at play.jobs.Job.call(Job.java:214)
	at play.jobs.Job$1.call(Job.java:119)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by gregorkistler** [comment]
I also get the former error message on each startup two times until the backend is completely up and reachable:
```
11:27:49,672 ERROR ~ Error executing job com.openseedbox.jobs.NodePollerJob$NodePollerWorker
com.openseedbox.code.MessageException: Unsuccessful webservice call, status: Not Found
```

But I didn't get the latter 🤔 Not via scheduler and also not when I run the job manually.
